rfcs/2020-03-06-1999-api-extensions-for-lua-transform.md

```diff
@@ -0,0 +1,175 @@
+
+## Guide-level Proposal
+
+### Motivating example
+
+
+```toml
+[transforms.lua]
+  type = "lua"
+  inputs = []
+  source = """
+    counter = counter + 1
+    event = nil
+  """
+  [transforms.lua.hooks]
+  init = """
+    counter = 0
+    previous_timestamp = os.time()
+    Event = Event.new_log()
+    event["message"] = "starting up"
+    event:set_lane("auxiliary")
+  """
+  shutdown = """
+    final_stats_event = Event.new_log()
+    final_stats_event["stats"] = { count = counter, interval = os.time() - previous_timestamp }
+    final_stats_event["stats.rate"] = final_stats_event["stats"].count / final_stats_event["stats.interval"]
+
+    shutdown_event = Event.new_log()
+    shutdown_event["message"] = "shutting down"
+    shutdown_event:set_lane("auxiliary")
+
+    event = {final_stats_event, shutdown_event}
+  """
+  [[transforms.lua.timers]]
+  interval = 10
+  source = """
+    event = Event.new_log()
+    event["stats"] = { count = counter, interval = 10 }
+    event["stats.rate"] = event["stats"].count / event["stats.interval"]
+    counter = 0
+    previous_timestamp = os.time()
+  """
+  [[transforms.lua.timers]]
+  interval = 60
+  source = """
+    event["message"] = "heartbeat"
+    event:set_lane("auxiliary")
+  ""
+```
+
+The code above consumes the incoming events, counts them, and then emits these stats about these counts every 10 seconds. In addition, it sends debug logs about its functioning into a separate lane called `auxiliary`.
+
+### Proposed changes
+
+* Hooks for initialization and shutdown called `init` and `shutdown`. They are defined as strings of Lua code in the `hooks` section of the configuration of the transform.
+* Timers which define pieces of code that are executed periodically. They are defined in array `timers`, each timer takes two configuration options: `interval` which is the interval for execution in seconds and `source` which is the code which is to be executed periodically.
+* Support for setting the output lane using `set_lane` method on the event which takes a string as the parameter. It should also be possible to read the lane using `get_lane` method. Reading from the lanes can be done in the downstream sinks by specifying the name of transform suffixed by a dot and the name of the lane.
+* Support multiple output events by making it possible to set the `event` global variable to an [sequence](https://www.lua.org/pil/11.1.html) of events.
+* Support direct access to the nested fields (in both maps and arrays).
+
+## Sales Pitch
+
+The proposal
+
+* gives users more power to create custom transforms;
+* does not break backward compatibility (except `pairs` method in case of nested fields);
+* makes it possible to add complexity to the configuration of the transform gradually only when needed.
+
+## Drawbacks
+
+The only drawback is that supporting both dot notation and classical indexing makes it impossible to add escaping of dots in field names. For example, for incoming event structure like
```

Comment:
That's a good point. I'm in favor of _not_ supporting the dot notation within runtimes, even if this is a breaking change. This syntax was one of the reasons that prompted a refactoring of our internal model, and dot notation should only be used in circumstances where a field must be accessed with a string (eg not in a runtime).